src/Microsoft.AspNet.Routing/Template/TemplateMatcher.cs

```diff
@@ -169,16 +169,53 @@ public class TemplateMatcher
                                          IReadOnlyDictionary<string, object> defaults,
                                          RouteValueDictionary values)
         {
+            var indexOfLastSegment = routeSegment.Parts.Count - 1;
+            if(routeSegment.Parts[indexOfLastSegment].IsOptional && 
+                routeSegment.Parts[indexOfLastSegment - 1].IsOptionalSeperator)
+            {
+                if (MatchComplexSegmentCore(routeSegment, requestSegment, Defaults, values, indexOfLastSegment))
+                {
+                    return true;
+                }
+                else
+                {
+                    if(requestSegment.EndsWith(routeSegment.Parts[indexOfLastSegment-1].Text))
+                    {
+                        return false;
+                    }
+                    
+                    return MatchComplexSegmentCore(
+                        routeSegment, 
+                        requestSegment, 
+                        Defaults, 
+                        values, 
+                        indexOfLastSegment - 2);
+                }
+            }
+            else
+            {
+                return MatchComplexSegmentCore(routeSegment, requestSegment, Defaults, values, indexOfLastSegment);
+            }
+        }
+
+        private bool MatchComplexSegmentCore(TemplateSegment routeSegment,
```

Comment:
The changes I have here are needed. indexOfLastSegmentUsed now comes from parameters. Also twe need to save the original values so if we return false we don't change them. It restore the new values if we are returning true.